### PR TITLE
VZ-3609 enable/disable component

### PR DESCRIPTION
**Pull request by guoyongzhang**
# Description

- add ApplicationOperatorComponent, OAMComponent, VerrazzanoComponent to ComponentSpec  in CRD
- add Enabled to IngressNginxComponent and IstioComponent in CRD
- remove IsEnabledFunc in HelmComponent and use IsEnabled in components
- implement IsEnabled for all components with Enabled field
- split install update in dynamic-updates tests into update1 and update2
- update getEffectiveCR() to add Certificate defaults for certManagerComponent

Fixes VZ-3609

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: ci/dynamic-updates/Jenkinsfile [hunk 1]
```diff
@@ -216,6 +216,10 @@ pipeline {
                             cd ${GO_REPO_PATH}/verrazzano
                             yq -i eval '.spec.components.weblogicOperator.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
                             yq -i eval '.spec.components.coherenceOperator.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
+                            # keycloak is not moved to go-based install yet
```
Comment (pfmackin) [inline ci/dynamic-updates/Jenkinsfile:219]: we should remove this keycloak commented lines and add it back when we do keyclaok

File: ci/dynamic-updates/Jenkinsfile [hunk 2]
```diff
@@ -268,10 +272,32 @@ pipeline {
                     }
                 }
 
-                stage('Post-Install Verify') {
+                // Update1 installs all components except COH/WLS
```
Comment (pfmackin) [inline ci/dynamic-updates/Jenkinsfile:275]: is this comment wrong?

File: ci/dynamic-updates/Jenkinsfile [hunk 3]
```diff
@@ -216,6 +216,10 @@ pipeline {
                             cd ${GO_REPO_PATH}/verrazzano
                             yq -i eval '.spec.components.weblogicOperator.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
                             yq -i eval '.spec.components.coherenceOperator.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
+                            # keycloak is not moved to go-based install yet
+                            # yq -i eval '.spec.components.keycloak.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
+                            yq -i eval '.spec.components.kiali.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
+                            yq -i eval '.spec.components.rancher.enabled = false' ${INSTALL_CONFIG_FILE_KIND}
                             ci/scripts/prepare_jenkins_at_environment.sh ${params.CREATE_CLUSTER_USE_CALICO} ${params.WILDCARD_DNS_DOMAIN}
```
Comment (pfmackin) [inline ci/dynamic-updates/Jenkinsfile:223]: Why do we have 2 update stages.  Why not install with everything disable that can be disabled, then update to enabled the components.

File: platform-operator/controllers/verrazzano/component/coherence/coherence.go [hunk 4]
```diff
@@ -22,8 +22,8 @@ func IsCoherenceOperatorReady(ctx spi.ComponentContext, _ string, namespace stri
 	return status.DeploymentsReady(ctx.Log(), ctx.Client(), deployments, 1)
 }
 
-// isCoherenceEnabled returns true if the component is enabled, which is the default
-func isCoherenceEnabled(ctx spi.ComponentContext) bool {
+// IsEnabled Coherence-specific enabled check for installation
+func (c coherenceComponent) IsEnabled(ctx spi.ComponentContext) bool {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/coherence/coherence.go:26]: nit-pick, this should probably be in `coherence_component.go`

File: ci/dynamic-updates/Jenkinsfile [hunk 5]
```diff
@@ -268,10 +272,32 @@ pipeline {
...
+                }
+
+                stage('Post-Update1 Verify Tests') {
                     environment {
                         TEST_ENV = "KIND"
-                        DUMP_DIRECTORY="${TEST_DUMP_ROOT}/post-install-verify-infra"
+                        DUMP_DIRECTORY="${TEST_DUMP_ROOT}/post-update1-verify-infra"
                     }
                     steps {
                         runGinkgoRandomize('verify-install')
```
Comment (michael-cico) [inline ci/dynamic-updates/Jenkinsfile:303]: Not part of this PR, but we probably need to modify the e2e tests to just check if a component is enabled to run the corresponding verify tests. 

Means though that we have to make the EffectiveCR or the merge logic available to the tests.

File: platform-operator/controllers/verrazzano/component/coherence/coherence.go (hunk 4 above)
Comment (guoyongzhang) [reply platform-operator/controllers/verrazzano/component/coherence/coherence.go]: yes.  Besides coherence, I also moved IsEnabled in WLS and keycloak.

File: ci/dynamic-updates/Jenkinsfile (hunk 5 above)
Comment (guoyongzhang) [reply ci/dynamic-updates/Jenkinsfile:299]: agreed.  In the future.

File: ci/dynamic-updates/Jenkinsfile (hunk 3 above)
Comment (guoyongzhang) [reply ci/dynamic-updates/Jenkinsfile:221]: The idea is 
1. install - with everything disabled 
2. update1 - enable everything except WLS/COH, run tests that don't need WLS/COH
3. update2 - enable WLS/COH, run all tests

For now, the step 1) installs only with keyloak/rancher/WLS/COH disabled.  Eventually, we will have all disabled.

File: ci/dynamic-updates/Jenkinsfile (hunk 2 above)
Comment (guoyongzhang) [reply ci/dynamic-updates/Jenkinsfile]: updated

File: ci/dynamic-updates/Jenkinsfile (hunk 1 above)
Comment (guoyongzhang) [reply ci/dynamic-updates/Jenkinsfile]: updated

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go [hunk 6]
```diff
@@ -43,5 +44,9 @@ func (e externalDNSComponent) IsReady(compContext spi.ComponentContext) bool {
 }
 
 func (e externalDNSComponent) IsEnabled(compContext spi.ComponentContext) bool {
-	return isEnabled(compContext)
+	dns := compContext.EffectiveCR().Spec.Components.DNS
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:47]: We should not do it this way.  We need to change the CR and add enabled to DNS.

Comment (guoyongzhang) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:47]: I guess the same goes for MySQL.  Can we get this merged and I will change these 2 in the next PR?

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:47]: are you saying we are adding `enabled` to the `DNSComponent` object?

What does it really mean if DNS is disabled?  if DNS is disabled by the flag, we still have to do wildcard right? seems like it might be confusing.

**Review by michael-cico — APPROVED** [review]
